sql: add sql statements and document fields of TABLE_STORAGE_STATS (#12777)
pingcap/docs: sql-statements/sql-statement-admin-cleanup.md

---
title: ADMIN CLEANUP
summary: An overview of the usage of ADMIN CLEANUP for the TiDB database.
---

# ADMIN CLEANUP INDEX

The `ADMIN CLEANUP INDEX` statement is used to delete redundant indexes from a table when the table has inconsistent data and index. Note that this syntax does not support [foreign key constraints](/foreign-key.md) yet.

## Synopsis

```ebnf+diagram
AdminCleanupStmt ::=
    'ADMIN' 'CLEANUP' ( 'INDEX' TableName IndexName | 'TABLE' 'LOCK' TableNameList )

TableNameList ::=
    TableName ( ',' TableName )*
```

## Examples

Assume that the `tbl` table in a database has inconsistent data and index due to some reasons (for example, some row data is lost in the cluster in a disaster recovery scenario):

```sql
SELECT * FROM tbl;
ERROR 1105 (HY000): inconsistent index idx handle count 3 isn't equal to value count 2

ADMIN CHECK INDEX tbl idx ;
ERROR 1105 (HY000): handle &kv.CommonHandle{encoded:[]uint8{0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf8}, colEndOffsets:[]uint16{0xa}}, index:types.Datum{k:0x5, decimal:0x0, length:0x0, i:0, collation:"utf8mb4_bin", b:[]uint8{0x0}, x:interface {}(nil)} != record:<nil>
```

> It can be seen from the error message of the `SELECT` query that, the `tbl` table contains 2 rows of data and 3 rows of index data, which means inconsistent row data and index data. At the same time, at least one index is in a dangling state. In this case, you can use the `ADMIN CLEANUP INDEX` statement to delete the dangling index:

Reviewer: ```suggestion
It can be seen from the error message of the `SELECT` query that, the `tbl` table contains two rows of data and three rows of index data, which means inconsistent row and index data. At the same time, at least one index is in a dangling state. In this case, you can use the `ADMIN CLEANUP INDEX` statement to delete the dangling index:
```

```sql
ADMIN CLEANUP INDEX tbl idx;
```

The execution result is as follows:

```sql
ADMIN CLEANUP INDEX tbl idx;
+---------------+
| REMOVED_COUNT |
+---------------+
|             1 |
+---------------+
```

> You can execute the `ADMIN CHECK INDEX` statement again to check the consistency of data and index, and verify whether the data has been restored to normal state:

Reviewer: ```suggestion
You can execute the `ADMIN CHECK INDEX` statement again to check the consistency of data and index, and verify whether the data has been restored to a normal state:
```

```sql
ADMIN CHECK INDEX tbl idx;
Query OK, 0 rows affected (0.01 sec)
```

### Precautions

* When the data and index are inconsistent due to the loss of replicas:

>     * There might be loss of both row data and index data. To address the issue, use the `ADMIN CLEANUP INDEX`and [`ADMIN RECOVER INDEX`](/sql-statements/sql-statement-admin-recover.md) statements together to restore the consistency of row data and index data.

Reviewer: ```suggestion
    * There might be a loss of both row data and index data. To restore the consistency, use the `ADMIN CLEANUP INDEX` and [`ADMIN RECOVER INDEX`](/sql-statements/sql-statement-admin-recover.md) statements together.
```

---

> title: ADMIN CLEANUP

Reviewer: ```suggestion
title: ADMIN CLEANUP INDEX
```

| REMOVED_COUNT |
+---------------+
|             1 |
+---------------+
```

You can execute the `ADMIN CHECK INDEX` statement again to check the consistency of data and index, and verify whether the data has been restored to a normal state:

```sql
ADMIN CHECK INDEX tbl idx;
Query OK, 0 rows affected (0.01 sec)
```

> **Note:**
>
> When the data and index are inconsistent due to the loss of replicas:
>
> - There might be a loss of both row data and index data. To restore the consistency, use the `ADMIN CLEANUP INDEX` and [`ADMIN RECOVER INDEX`](/sql-statements/sql-statement-admin-recover.md) statements together.
> - The `ADMIN CLEANUP INDEX` statement is always executed in a single thread. When the table data is large, it is recommended to recover the index data by rebuilding the index.
> - When you execute the `ADMIN CLEANUP INDEX` statement, the corresponding table or index is not locked and TiDB allows other sessions to modify the table records at the same time. However, in this case, `ADMIN CLEANUP INDEX` might not be able to handle all table records correctly. Therefore, when you execute `ADMIN CLEANUP INDEX`, avoid modifying the table data at the same time.

> > - If you use the enterprise edition of TiDB, you can contact the support engineer for help.

Author: ```suggestion
> - If you use the enterprise edition of TiDB, you can [submit a request](https://support.pingcap.com/hc/en-us) to contact the support engineer for help.
```
